drivers/retained_mem/retained_mem_zephyr_bbram.c

```diff
@@ -0,0 +1,164 @@
+/*
+ * Copyright (c) 2025, Nordic Semiconductor ASA, RBR Ltd
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_retained_bbram
+
+#include <string.h>
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/bbram.h>
+#include <zephyr/drivers/retained_mem.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retained_mem_zephyr_bbram, CONFIG_RETAINED_MEM_LOG_LEVEL);
+
+#ifdef CONFIG_RETAINED_MEM_MUTEXES
+struct retained_bbram_data {
+	struct k_mutex lock;
+};
+#endif
+
+struct retained_bbram_config {
+	const struct device *bbram;
+};
+
+static inline void retained_bbram_lock(const struct device *dev)
+{
+#ifdef CONFIG_RETAINED_MEM_MUTEXES
+	struct retained_bbram_data *data = dev->data;
+
+	k_mutex_lock(&data->lock, K_FOREVER);
+#else
+	ARG_UNUSED(dev);
+#endif
+}
+
+static inline void retained_bbram_unlock(const struct device *dev)
+{
+#ifdef CONFIG_RETAINED_MEM_MUTEXES
+	struct retained_bbram_data *data = dev->data;
+
+	k_mutex_unlock(&data->lock);
+#else
+	ARG_UNUSED(dev);
+#endif
+}
+
+static int retained_bbram_init(const struct device *dev)
+{
+#ifdef CONFIG_RETAINED_MEM_MUTEXES
+	struct retained_bbram_data *data = dev->data;
+
+	k_mutex_init(&data->lock);
+#else
+	ARG_UNUSED(dev);
+#endif
+
+	return 0;
+}
+
+static ssize_t retained_bbram_size(const struct device *dev)
+{
+	const struct retained_bbram_config *config = dev->config;
+	int err;
+	size_t size;
+
+	err = bbram_get_size(config->bbram, &size);
+	if (err < 0) {
+		return (ssize_t)err;
+	}
+
+	return (ssize_t)size;
+}
+
+static int retained_bbram_read(const struct device *dev, off_t offset, uint8_t *buffer, size_t size)
+{
+	const struct retained_bbram_config *config = dev->config;
+	int err;
+
+	retained_bbram_lock(dev);
+
+	err = bbram_read(config->bbram, offset, size, buffer);
+
+	retained_bbram_unlock(dev);
+
+	return err;
+}
+
+static int retained_bbram_write(const struct device *dev, off_t offset, const uint8_t *buffer,
+				size_t size)
+{
+	const struct retained_bbram_config *config = dev->config;
+
+	retained_bbram_lock(dev);
+
+	int err = bbram_write(config->bbram, offset, size, buffer);
+
+	retained_bbram_unlock(dev);
+
+	return err;
+}
+
+static int retained_bbram_clear(const struct device *dev)
+{
+	const struct retained_bbram_config *config = dev->config;
+	uint8_t clear[CONFIG_RETAINED_MEM_ZEPHYR_BBRAM_CLEAR_CHUNK_SIZE];
+	size_t total_size;
+	size_t chunk_size = CONFIG_RETAINED_MEM_ZEPHYR_BBRAM_CLEAR_CHUNK_SIZE;
+	int err = 0;
+
+	memset(clear, 0, CONFIG_RETAINED_MEM_ZEPHYR_BBRAM_CLEAR_CHUNK_SIZE);
+
+	retained_bbram_lock(dev);
+
+	err = bbram_get_size(config->bbram, &total_size);
+	if (err < 0) {
+		return err;
```

Comment:
You should either unlock here or use a goto and have a single exit point for the function.